lib/src/kbasesearchengine/events/storage/MongoDBStatusEventStorage.java

```diff
@@ -317,4 +317,24 @@ public boolean setProcessingState(
         return Optional.of(toEvent(ret));
     }
 
+    @Override
```

Comment:
Here I think some discussion is warranted. From an implementation standpoint, IMO this should just be done with a mongo query that finds events with FAIL and sets them to UNPROC.

The bigger issue is whether resetting all events makes sense. There's probably some events that are never going to pass and that should not be rerun. I haven't completely thought this through yet, but there may be events that are dangerous to rerun. Thoughts?

Reply:
I am updating the method to use a mongo query. Given that we are now seeking the latest state of the storage objects, I don't see how anything can go wrong with just this function of correctly indexing objects with out-of-order events. There may be other unrelated issues that constantly prevent the event from being updated or the object from being indexed. For this, we could have another CLI to delete failed events that are older than a certain threshold timestamp. I would imagine that this would be done after logs have been reviewed, an attempt to fix problems has been made and a decision to purge these events is the option of last resort.

Reply:
Fair enough. If there are events that turn out to be dangerous we can deal with them as they come up.

Rather than delete the events we could just  move them to a permanently failed state.

I'm expecting that we'll need some sort of UI to change states on selected events and allow more state transitions (I'm particularly thinking of PROC->UNPROC) but we can do that with mongo surgery for now.

Reply:
Been thinking about this a bit more, and realized there are some events that in some cases you don't want to run again because some of the state is reflected in the event type itself, i.e.

PUBLISH_ACCESS_GROUP and UNPUBLISH_ACCESS_GROUP
Since the event type determines what state is set on the records in ES, running a failed PAG after a UAG would put the records in an inconsistent state with the WS.

DELETE_ALL_VERSIONS and UNDELETE_ALL_VERSIONS
Same logic as above

DELETE_ACCESS_GROUP and UNDELETE_ACCESS_GROUP
yadda yadda

The delete events aren't a huge problem because
1) the WS doesn't emit UNDELETE_ACCESS_GROUP events
2) The UI doesn't support undelete operations

I think the other events are safe to rerun.

Obviously this again argues for removing state from the events, or at least collapsing the events in some way such that the state can be easily updated like with RENAME_OBJECT, for example.

Reply:
How about, 
- If the WS object is deleted and index contains the object, delete object in index
- else, delete object in index and treat failed event as a new version event

Reply:
Do you mean delete as in remove from the index? If so I don't think there's a need to do that; marking it deleted is fine.

IMO we shouldn't worry about deletion for now since undelete events aren't implemented by the UI. We should, however, note all the problems with deletion and undeletion (e.g. undelete needs to update the state of the index from the ws since object state may have changed since the last index, and out of order deletions/undeletions) in implementation_notes.md

There does need to be a solution for the publish events though. Maybe it's worth it to collapse those into one event so that the state can be updated.